File: gemini_ts/vertex-ai-tool.ts
import { VertexAI, GenerativeModel, Part } from '@google-cloud/vertexai';
import * as fs from 'fs';
import * as path from 'path';
import * as dotenv from 'dotenv';
import { fileURLToPath } from 'url';
import { ProxyAgent, setGlobalDispatcher } from 'undici';

// Load environment variables
dotenv.config();
// Try loading from parent directory if not found or specific vars are missing
if (!process.env.VERTEX_PROJECT_ID && !process.env.GOOGLE_APPLICATION_CREDENTIALS) {
  const parentEnvPath = path.resolve(process.cwd(), '..', '.env');
  if (fs.existsSync(parentEnvPath)) {
    dotenv.config({ path: parentEnvPath });
  }
}

// Configure proxy if available
const proxyUrl = process.env.HTTPS_PROXY || process.env.HTTP_PROXY;
if (proxyUrl) {
  const dispatcher = new ProxyAgent(proxyUrl);
  setGlobalDispatcher(dispatcher);
  console.log(`Using proxy: ${proxyUrl}`);
}

/**
 * 独立的 Vertex AI 调用工具类
 * 支持文本大模型和图像大模型的调用
 */
export class VertexAITool {
  private project_id: string;
  private location: string;
  private credentials_path: string | undefined;
  private text_model_name: string;
  private image_model_name: string;
  private timeout: number;
  private max_retries: number;
  private vertex_ai: VertexAI;
  private text_model: GenerativeModel;
  private image_model: GenerativeModel;

  /**
   * 初始化 Vertex AI 工具类
   * 
   * @param config 配置对象
   */
  constructor(config: {
    project_id?: string;
    location?: string;
    credentials_path?: string;
    text_model?: string;
    image_model?: string;
    timeout?: number;
    max_retries?: number;
  } = {}) {
    this.project_id = config.project_id || process.env.VERTEX_PROJECT_ID || '';
    this.location = config.location || process.env.VERTEX_LOCATION || 'us-central1';
    this.credentials_path = config.credentials_path || process.env.GOOGLE_APPLICATION_CREDENTIALS;
    this.text_model_name = config.text_model || process.env.TEXT_MODEL || 'gemini-3-flash-preview';
    this.image_model_name = config.image_model || process.env.IMAGE_MODEL || 'gemini-3-pro-image-preview';
    this.timeout = config.timeout || 300;
    this.max_retries = config.max_retries || 2;

    // 自动纠正不支持的 location
    if (this.location === 'global') {
      console.warn("⚠️ Warning: 'global' location is not supported for Vertex AI GenAI models. Automatically switching to 'us-central1'.");
      this.location = 'us-central1';
    }

    // 打印调试信息
    console.log("--- Vertex AI Configuration ---");
    console.log(`Project ID: ${this.project_id}`);
    console.log(`Location: ${this.location}`);
    console.log(`Credentials: ${this.credentials_path || 'Not set (using default/ADC)'}`);
    console.log(`Text Model: ${this.text_model_name}`);
    console.log(`Image Model: ${this.image_model_name}`);
    console.log("-------------------------------");

    if (!this.project_id) {
      console.warn("⚠️ Warning: Project ID is not set. API calls will likely fail.");
      console.warn("Please set VERTEX_PROJECT_ID in your .env file.");
    }

    if (this.credentials_path) {
      process.env.GOOGLE_APPLICATION_CREDENTIALS = this.credentials_path;
    }

    // 初始化 Vertex AI Client
    this.vertex_ai = new VertexAI({
      project: this.project_id,
      location: this.location,
    });

    // 初始化模型实例
    this.text_model = this.vertex_ai.getGenerativeModel({
      model: this.text_model_name,
    });
    
    this.image_model = this.vertex_ai.getGenerativeModel({
      model: this.image_model_name,
    });
  }

  /**
   * 简单的重试包装器
   */
  private async withRetry<T>(fn: () => Promise<T>, retries: number = 3): Promise<T> {
    let lastError: any;
    for (let i = 0; i < retries; i++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error;
        const delay = Math.pow(2, i) * 1000;
        console.warn(`Attempt ${i + 1} failed, retrying in ${delay}ms...`, error);
        await new Promise(resolve => setTimeout(resolve, delay));
      }
    }
    throw lastError;
  }

  /**
   * 使用 Vertex AI 文本模型生成文本
   * 
   * @param prompt 输入提示词
   * @param thinking_budget 推理预算（0 表示禁用推理模式）
   * @returns 生成的文本
   */
  async generateText(prompt: string, thinking_budget: number = 0): Promise<string> {
    return this.withRetry(async () => {
      // 当前 SDK 可能还不支持 thinkingConfig，但我们可以尝试作为 generationConfig 传入
      // 或者如果不支持，就忽略它。
      // 注意：Node SDK 对于特定 beta 功能的支持可能滞后。
      // 这里暂时只传 prompt。
      
      const request: any = {
        contents: [{ role: 'user', parts: [{ text: prompt }] }]
      };

      if (thinking_budget > 0) {
        // 尝试添加 thinking_config (如果 SDK 支持的话，否则可能会被忽略或报错)
        // 目前看来 Node SDK 的 generationConfig 类型中还没有 thinkingConfig
        // 我们暂时先跳过 thinking_budget 的设置，以免出错
        console.warn("Thinking budget is not yet fully supported in this Node.js SDK wrapper.");
      }

      const result = await this.text_model.generateContent(request);
      const response = await result.response;
      
      if (!response.candidates || response.candidates.length === 0) {
        throw new Error("No candidates returned");
      }
      
      const textPart = response.candidates[0].content.parts.find(p => p.text);
      return textPart ? textPart.text || '' : '';
    });
  }

  /**
   * 使用 Vertex AI 文本模型生成文本（支持图像输入）
   * 
   * @param prompt 输入提示词
   * @param imagePath 图像文件路径
   * @param thinking_budget 推理预算（0 表示禁用推理模式）
   * @returns 生成的文本
   */
  async generateWithImage(prompt: string, imagePath: string, thinking_budget: number = 0): Promise<string> {
    return this.withRetry(async () => {
      const imageBuffer = fs.readFileSync(imagePath);
      const base64Image = imageBuffer.toString('base64');
      const mimeType = this.getMimeType(imagePath);

      const request: any = {
        contents: [{
          role: 'user',
          parts: [
            { inlineData: { data: base64Image, mimeType: mimeType } },
            { text: prompt }
          ]
        }]
      };

      const result = await this.text_model.generateContent(request);
      const response = await result.response;

      if (!response.candidates || response.candidates.length === 0) {
        throw new Error("No candidates returned");
      }

      const textPart = response.candidates[0].content.parts.find(p => p.text);
      return textPart ? textPart.text || '' : '';
    });
  }

  /**
   * 使用 Vertex AI 图像模型生成图像
   * 
   * @param prompt 图像生成提示词
   * @param refImages 参考图像路径列表 (Buffer 或 Base64)
   * @param aspect_ratio 图像比例
   * @param resolution 图像分辨率
   * @param enable_thinking 是否启用推理模式
   * @param thinking_budget 推理预算
   * @returns 生成的图像 Buffer，失败返回 null
   */
  async generateImage(
    prompt: string,
    refImages?: string[],
    aspect_ratio: string = "16:9",
    resolution: string = "2K",
    enable_thinking: boolean = true,
    thinking_budget: number = 1024
  ): Promise<Buffer | null> {
    return this.withRetry(async () => {
      try {
        const parts: Part[] = [];

        if (refImages) {
          for (const imgPath of refImages) {
            const imageBuffer = fs.readFileSync(imgPath);
            parts.push({
              inlineData: {
                data: imageBuffer.toString('base64'),
                mimeType: this.getMimeType(imgPath)
              }
            });
          }
        }

        parts.push({ text: prompt });

        // 构建请求配置
        // 注意：Node SDK 中 imageConfig 的支持情况可能与 Python 不同
        // 这里尝试构造类似 Python 的配置结构，但可能需要调整
        const generationConfig: any = {
           // 这里我们假设 SDK 会透传这些配置
           // 但实际上 @google-cloud/vertexai 的类型定义可能比较严格
           // 如果 TS 报错，可能需要忽略类型检查
        };

        // 由于 GenerateContentRequest 的 generationConfig 类型限制，我们可能无法直接传入 imageConfig
        // 如果是 Imagen 模型，应该使用 image generation API 而不是 generateContent
        // 但既然 Python 代码用了 generateContent，我们这里也尝试用它。
        
        // 构建请求体
        const request: any = {
          contents: [{ role: 'user', parts: parts }],
          generationConfig: generationConfig
        };

        const result = await this.image_model.generateContent(request);
        const response = await result.response;

        if (!response.candidates || response.candidates.length === 0) {
          throw new Error("No candidates returned");
        }
        
        // 查找图像部分
        for (const candidate of response.candidates) {
            for (const part of candidate.content.parts) {
                // 检查 inlineData 是否存在且是图像
                if (part.inlineData && part.inlineData.mimeType && part.inlineData.mimeType.startsWith('image/')) {
                    return Buffer.from(part.inlineData.data, 'base64');
                }
            }
        }

        throw new Error("响应中未找到图像");
      } catch (error) {
        console.error(`图像生成失败: ${error}`);
        throw error;
      }
    });
  }

  private getMimeType(filePath: string): string {
    const ext = path.extname(filePath).toLowerCase();
    switch (ext) {
      case '.png': return 'image/png';
      case '.jpg':
      case '.jpeg': return 'image/jpeg';
      case '.webp': return 'image/webp';
      case '.gif': return 'image/gif';
      default: return 'image/png';
    }
  }
}

// 测试代码
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const main = async () => {
    const vertexTool = new VertexAITool();
    console.log("=".repeat(60));
    console.log("Vertex AI 工具类测试 (TypeScript)");
    console.log("=".repeat(60));

    const readline = await import('readline');
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });

    const question = (query: string) => new Promise<string>(resolve => rl.question(query, resolve));

    while (true) {
      console.log("\n请选择测试类型:");
      console.log("1. 文本模型测试");
      console.log("2. 图像模型测试");
      console.log("3. 退出");
      const choice = (await question("请输入选项 (1-3): ")).trim();

      if (choice === "1") {
        console.log("\n--- 文本模型测试 ---");
        let prompt = (await question("请输入提示词 (默认: '你好，请简单介绍一下你自己。'): ")).trim();
        if (!prompt) prompt = "你好，请简单介绍一下你自己。";

        try {
          console.log("\n正在调用 Vertex AI 文本模型...");
          const result = await vertexTool.generateText(prompt);
          console.log("\n生成结果:");
          console.log("-".repeat(60));
          console.log(result);
          console.log("-".repeat(60));
        } catch (e) {
          console.error(`\n错误: ${e}`);
        }
      } else if (choice === "2") {
        console.log("\n--- 图像模型测试 ---");
        let prompt = (await question("请输入图像生成提示词 (默认: '一只可爱的猫咪在花园里玩耍'): ")).trim();
        if (!prompt) prompt = "一只可爱的猫咪在花园里玩耍";

        let savePath = (await question("请输入保存路径 (默认: 'generated_image_ts.png'): ")).trim();
        if (!savePath) savePath = "generated_image_ts.png";

        try {
          console.log("\n正在调用 Vertex AI 图像模型...");
          console.log(`提示词: ${prompt}`);
          const result = await vertexTool.generateImage(prompt);

          if (result) {
            fs.writeFileSync(savePath, result);
            console.log(`\n图像已保存到: ${savePath}`);
          } else {
            console.log("\n图像生成失败");
          }
        } catch (e) {
          console.error(`\n错误: ${e}`);
        }
      } else if (choice === "3") {
        console.log("\n退出测试");
        rl.close();
        break;
      } else {
        console.log("\n无效选项，请重新选择");
      }
    }
  };

  main().catch(console.error);
}

File: vertex_ai_tool.py
"""
独立的 Vertex AI 调用工具类
支持文本大模型和图像大模型的调用
"""
import os
import logging
from typing import Optional, List
from google import genai
from google.genai import types
from PIL import Image
from io import BytesIO
from tenacity import retry, stop_after_attempt, wait_exponential
from dotenv import load_dotenv

logger = logging.getLogger(__name__)


class VertexAITool:
    """Vertex AI 调用工具类，整合文本和图像大模型"""

    def __init__(
        self,
        project_id: str = None,
        location: str = None,
        credentials_path: str = None,
        text_model: str = "gemini-3.1-pro-preview",
        image_model: str = "gemini-3.1-flash-image-preview",
        timeout: float = 300.0,
        max_retries: int = 2
    ):
        """
        初始化 Vertex AI 工具类

        Args:
            project_id: GCP 项目 ID
            location: GCP 区域
            credentials_path: 服务账号凭据文件路径
            text_model: 文本模型名称
            image_model: 图像模型名称
            timeout: 超时时间（秒）
            max_retries: 最大重试次数
        """
        self.project_id = project_id
        self.location = location or 'us-central1'
        self.credentials_path = credentials_path
        self.text_model = text_model
        self.image_model = image_model
        self.timeout = timeout
        self.max_retries = max_retries

        if credentials_path:
            os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = credentials_path

        timeout_ms = int(self.timeout * 1000)
        self.client = genai.Client(
            vertexai=True,
            project=self.project_id,
            location=self.location,
            http_options=types.HttpOptions(timeout=timeout_ms)
        )

    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=2, max=10),
        reraise=True
    )
    def generate_text(self, prompt: str, thinking_budget: int = 0) -> str:
        """
        使用 Vertex AI 文本模型生成文本

        Args:
            prompt: 输入提示词
            thinking_budget: 推理预算（0 表示禁用推理模式）

        Returns:
            生成的文本
        """
        config_params = {}
        if thinking_budget > 0:
            config_params['thinking_config'] = types.ThinkingConfig(thinking_budget=thinking_budget)

        response = self.client.models.generate_content(
            model=self.text_model,
            contents=prompt,
            config=types.GenerateContentConfig(**config_params) if config_params else None,
        )
        return response.text

    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=2, max=10),
        reraise=True
    )
    def generate_with_image(self, prompt: str, image_path: str, thinking_budget: int = 0) -> str:
        """
        使用 Vertex AI 文本模型生成文本（支持图像输入）

        Args:
            prompt: 输入提示词
            image_path: 图像文件路径
            thinking_budget: 推理预算（0 表示禁用推理模式）

        Returns:
            生成的文本
        """
        img = Image.open(image_path)
        contents = [img, prompt]

        config_params = {}
        if thinking_budget > 0:
            config_params['thinking_config'] = types.ThinkingConfig(thinking_budget=thinking_budget)

        response = self.client.models.generate_content(
            model=self.text_model,
            contents=contents,
            config=types.GenerateContentConfig(**config_params) if config_params else None,
        )
        return response.text

    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=2, min=4, max=60),
        reraise=True
    )
    def generate_image(
        self,
        prompt: str,
        ref_images: Optional[List[Image.Image]] = None,
        aspect_ratio: str = "16:9",
        resolution: str = "2K",
        enable_thinking: bool = True,
        thinking_budget: int = 1024
    ) -> Optional[Image.Image]:
        """
        使用 Vertex AI 图像模型生成图像

        Args:
            prompt: 图像生成提示词
            ref_images: 参考图像列表
            aspect_ratio: 图像比例
            resolution: 图像分辨率
            enable_thinking: 是否启用推理模式
            thinking_budget: 推理预算

        Returns:
            生成的 PIL Image 对象，失败返回 None
        """
        try:
            contents = []

            if ref_images:
                for ref_img in ref_images:
                    contents.append(ref_img)

            contents.append(prompt)

            config_params = {
                'response_modalities': ['TEXT', 'IMAGE'],
                'image_config': types.ImageConfig(
                    aspect_ratio=aspect_ratio,
                    image_size=resolution
                )
            }

            if enable_thinking:
                config_params['thinking_config'] = types.ThinkingConfig(
                    thinking_budget=thinking_budget,
                    include_thoughts=True
                )

            response = self.client.models.generate_content(
                model=self.image_model,
                contents=contents,
                config=types.GenerateContentConfig(**config_params)
            )

            last_image = None
            for i, part in enumerate(response.parts):
                if part.text is None:
                    try:
                        image = part.as_image()
                        if image:
                            if isinstance(image, Image.Image):
                                last_image = image
                            elif hasattr(image, 'image_bytes') and image.image_bytes:
                                last_image = Image.open(BytesIO(image.image_bytes))
                            elif hasattr(image, '_pil_image') and image._pil_image:
                                last_image = image._pil_image
                    except Exception as e:
                        logger.warning(f"提取图像失败: {e}")

            if last_image:
                return last_image

            raise ValueError("响应中未找到图像")

        except Exception as e:
            logger.error(f"图像生成失败: {e}", exc_info=True)
            raise Exception(f"图像生成失败: {e}") from e


if __name__ == "__main__":
    """简单的测试代码，测试 Vertex AI 文本和图像大模型调用"""
    import sys

    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
    )

    load_dotenv()

    vertex_tool = VertexAITool(
        project_id=os.getenv('VERTEX_PROJECT_ID'),
        location=os.getenv('VERTEX_LOCATION'),
        credentials_path=os.getenv('GOOGLE_APPLICATION_CREDENTIALS'),
        text_model=os.getenv('TEXT_MODEL', 'gemini-3-flash-preview'),
        image_model=os.getenv('IMAGE_MODEL', 'gemini-3-pro-image-preview')
    )

    print("=" * 60)
    print("Vertex AI 工具类测试")
    print("=" * 60)

    test_image_path = os.path.join(os.path.dirname(__file__), 'assets', 'test_img.png')

    while True:
        print("\n请选择测试类型:")
        print("1. 文本模型测试")
        print("2. 图像模型测试")
        print("3. 退出")
        choice = input("请输入选项 (1-3): ").strip()

        if choice == "1":
            print("\n--- 文本模型测试 ---")
            prompt = input("请输入提示词 (默认: '你好，请简单介绍一下你自己。'): ").strip()
            if not prompt:
                prompt = "你好，请简单介绍一下你自己。"

            try:
                print(f"\n正在调用 Vertex AI 文本模型...")
                result = vertex_tool.generate_text(prompt)
                print("\n生成结果:")
                print("-" * 60)
                print(result)
                print("-" * 60)
            except Exception as e:
                print(f"\n错误: {e}")

        elif choice == "2":
            print("\n--- 图像模型测试 ---")
            prompt = input("请输入图像生成提示词 (默认: '一只可爱的猫咪在花园里玩耍'): ").strip()
            if not prompt:
                prompt = "一只可爱的猫咪在花园里玩耍"

            save_path = input("请输入保存路径 (默认: 'generated_image.png'): ").strip()
            if not save_path:
                save_path = "generated_image.png"

            try:
                print(f"\n正在调用 Vertex AI 图像模型...")
                print(f"提示词: {prompt}")
                result = vertex_tool.generate_image(prompt)

                if result:
                    result.save(save_path)
                    print(f"\n图像已保存到: {save_path}")
                    print(f"图像尺寸: {result.size}")
                else:
                    print("\n图像生成失败")
            except Exception as e:
                print(f"\n错误: {e}")

        elif choice == "3":
            print("\n退出测试")
            break

        else:
            print("\n无效选项，请重新选择")
